File: services/backend/src/deskops_backend/domain/telemetry_schema.py
from deskops_backend.domain.models import TelemetryEvent, is_valid_iso8601


def validate_telemetry_event(event: TelemetryEvent) -> tuple[bool, str | None]:
    if not event.event_id.strip():
        return False, "missing_or_invalid_event_id"
    if not event.category.strip():
        return False, "missing_or_invalid_category"
    if event.result not in {"success", "failed", "review"}:
        return False, "invalid_result"
    if not is_valid_iso8601(event.occurred_at):
        return False, "invalid_occurred_at"
    return True, None

File: archive/legacy-node-backend/src/domain/telemetry-schema.mjs
export function validateTelemetryEvent(event) {
  const required = ["eventId", "occurredAt", "category", "result"];
  for (const key of required) {
    if (!event || typeof event[key] !== "string" || event[key].trim() === "") {
      return { valid: false, reason: `missing_or_invalid_${key}` };
    }
  }

  if (!["success", "failed", "review"].includes(event.result)) {
    return { valid: false, reason: "invalid_result" };
  }

  if (Number.isNaN(new Date(event.occurredAt).getTime())) {
    return { valid: false, reason: "invalid_occurredAt" };
  }

  return { valid: true };
}
